### PR TITLE
Create Solana Donations Integration

### PR DESCRIPTION
**Pull request by CarlosQ96**
Keeping the integration as simple as posible and backward compatible. 

Need to update solana token I investigated and found different ids from Alquemy and Covalent websites, but I am not convinced with those values.

### CODE REVIEW

File: migration/1703044586989-addSolanaToken.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+export class addSolanaToken1703044586989 implements MigrationInterface {
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.manager.save(
+      Token,
+      seedTokens
+        .filter(token => token.networkId === NETWORK_IDS.SOLANA)
+        .map(t => {
+          t.address = t.address?.toLowerCase();
+          return t;
```
Comment (aminlatifi) [inline migration/1703044586989-addSolanaToken.ts]: The filter must be based on the `chainType` rather than `networkId`. Plus, address capitalization is important in Solana and must be preserved.

Comment (aminlatifi) [reply migration/1703044586989-addSolanaToken.ts]: Though, at this stage we will only add solana native token, so maybe it's better to apply more strict filtering conditions.

File: migration/1703044586989-addSolanaToken.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+                  (${token.id}, ${givethOrganization.id}),
+                  (${token.id}, ${traceOrganization.id})
+                ;`);
+    }
+  }
+
+  public async down(queryRunner: QueryRunner): Promise<void> {
+    const tokens = await queryRunner.query(`
+            SELECT * FROM token
+            WHERE "networkId" = ${NETWORK_IDS.SOLANA}
```
Comment (aminlatifi) [inline migration/1703044586989-addSolanaToken.ts]: The same as above in reverting... chainType and native solana token address must be the criteria

File: migration/data/seedTokens.ts [hunk 3]
```diff
@@ -1247,6 +1249,15 @@ const seedTokens: ITokenData[] = [
     networkId: NETWORK_IDS.MORDOR_ETC_TESTNET,
   },
   // TODO: Add solana token
+  {
+    name: 'Solana native token',
+    symbol: 'SOL',
+    address: '0x0000000000000000000000000000000000000000',
```
Comment (aminlatifi) [inline migration/data/seedTokens.ts]: Solana native token doesn't have any address, it's also mentioned https://github.com/Giveth/impact-graph/issues/1175

But we we can rely on system **programId** `11111111111111111111111111111111` which is involved in native solana token transfer as well.

File: migration/data/seedTokens.ts [hunk 4]
```diff
@@ -1247,6 +1249,15 @@ const seedTokens: ITokenData[] = [
...
+    decimals: 18,
```
Comment (aminlatifi) [inline migration/data/seedTokens.ts]: that one SOL can be divided into 1,000,000,000 (one billion) smaller units. These smallest units are often referred to as lamports
```suggestion
    decimals: 9,
```

File: src/adapters/price/MonoswapPriceAdapter.ts [hunk 5]
```diff
@@ -26,7 +26,7 @@ export class MonoswapPriceAdapter implements PriceAdapterInterface {
     const tokenPrices = await getMonoSwapTokenPrices(
       params.symbol,
       baseTokens,
-      params.networkId,
+      params.networkId!,
```
Comment (aminlatifi) [inline src/adapters/price/MonoswapPriceAdapter.ts]: Is it necessary?

File: src/adapters/price/PriceAdapterInterface.ts [hunk 6]
```diff
@@ -1,6 +1,6 @@
 export interface GetTokenPriceParams {
   symbol: string;
-  networkId: number;
+  networkId?: number;
```
Comment (aminlatifi) [inline src/adapters/price/PriceAdapterInterface.ts]: Is it necessary?

Comment (aminlatifi) [reply src/adapters/price/PriceAdapterInterface.ts]: Let's keep it mandatory and use 0 as networkId on solana, the code will be more type safe.

File: migration/data/seedTokens.ts [hunk 7]
```diff
@@ -1247,6 +1249,15 @@ const seedTokens: ITokenData[] = [
...
+    networkId: 0,
```
Comment (aminlatifi) [inline migration/data/seedTokens.ts]:
```suggestion
    networkId: NETWORK_IDS.SOLANA,,
```

File: src/resolvers/donationResolver.test.ts [hunk 8]
```diff
@@ -817,6 +819,53 @@ function createDonationTestCases() {
...
+
+    const user = await User.create({
+      walletAddress: generateRandomSolanaAddress(),
+      loginType: 'wallet',
+      firstName: 'first name',
+    }).save();
+
+    const referredEvent = await firstOrCreateReferredEventByUserId(user.id);
+    referredEvent.startTime = new Date();
+    await referredEvent.save();
```
Comment (aminlatifi) [inline src/resolvers/donationResolver.test.ts]: Do we need this? it is probably copied from a QF-related donation test.

File: src/services/authorizationServices.ts [hunk 9]
```diff
@@ -77,6 +80,7 @@ export const validateImpactGraphJwt = async (
 ): Promise<JwtVerifiedUser> => {
   const secret = config.get('JWT_SECRET') as string;
   const decodedJwt: any = jwt.verify(token, secret);
+  const chainType = detectAddressChainType(decodedJwt?.walletAddress);
```
Comment (aminlatifi) [inline src/services/authorizationServices.ts]: It's not efficient to do this detection and fill chainType for every authentication. Let's leave it as it is an only do the detection in required points (e.g. donation resolver)

File: src/services/authorizationServices.ts [hunk 10]
```diff
@@ -49,6 +51,7 @@ export interface JwtVerifiedUser {
   lastName?: string;
   walletAddress?: string;
   userId: number;
+  chainType: string;
```
Comment (aminlatifi) [inline src/services/authorizationServices.ts]: This must be removed so

File: src/services/authorizationServices.ts [hunk 11]
```diff
@@ -85,6 +89,7 @@ export const validateImpactGraphJwt = async (
     lastName: decodedJwt?.lastName,
     walletAddress: decodedJwt?.walletAddress,
     userId: decodedJwt?.userId,
+    chainType: chainType!,
```
Comment (aminlatifi) [inline src/services/authorizationServices.ts]: This must be removed so

File: src/services/authorizationServices.ts [hunk 12]
```diff
@@ -116,13 +121,16 @@ export const validateAuthMicroserviceJwt = async (
       user = await createUserWithPublicAddress(userAddress);
     }
 
+    const chainType = detectAddressChainType(user?.walletAddress);
+
```
Comment (aminlatifi) [inline src/services/authorizationServices.ts]: Like the other, it must be removed.

File: src/services/authorizationServices.ts [hunk 13]
```diff
@@ -116,13 +121,16 @@ export const validateAuthMicroserviceJwt = async (
...
     return {
       email: user?.email,
       firstName: user?.firstName,
       lastName: user?.lastName,
       name: user?.name,
       walletAddress: user?.walletAddress,
       userId: user!.id,
+      chainType: chainType!,
```
Comment (aminlatifi) [inline src/services/authorizationServices.ts]: Remove this one

File: src/resolvers/donationResolver.ts [hunk 14]
```diff
@@ -626,8 +627,9 @@ export class DonationResolver {
   ): Promise<Number> {
     try {
       const userId = ctx?.req?.user?.userId;
+      const chainType = ctx?.req?.user?.chainType;
       const donorUser = await findUserById(userId);
-      if (!donorUser) {
+      if (!chainType || !donorUser) {
```
Comment (aminlatifi) [inline src/resolvers/donationResolver.ts]: As I mentioned in another comment, it should not be expected to jwt data contain chainType. We can do the chain-type detection here.

File: src/resolvers/donationResolver.ts [hunk 15]
```diff
@@ -647,25 +649,6 @@ export class DonationResolver {
...
-          priceChainId = NETWORK_IDS.OPTIMISTIC;
-          break;
-        case NETWORK_IDS.MORDOR_ETC_TESTNET:
-          priceChainId = NETWORK_IDS.ETC;
-          break;
-        default:
-          priceChainId = transactionNetworkId;
-          break;
-      }
-
```
Comment (aminlatifi) [inline src/resolvers/donationResolver.ts]: Some logics from here are removed! they seem important

Comment (aminlatifi) [reply src/resolvers/donationResolver.ts]: This snippet can be updated to support solana as well

File: src/resolvers/donationResolver.ts [hunk 16]
```diff
@@ -770,14 +754,36 @@ export class DonationResolver {
...
       await donation.save();
 
-      await updateDonationPricesAndValues(
-        donation,
-        project,
-        tokenInDb,
-        token,
-        priceChainId,
-        amount,
-      );
```
Comment (aminlatifi) [inline src/resolvers/donationResolver.ts]: And then keep it as it simple is

File: src/resolvers/donationResolver.test.ts [hunk 17]
```diff
@@ -817,6 +819,53 @@ function createDonationTestCases() {
...
+        },
+      },
+    );
+    assert.isOk(saveDonationResponse.data.data.createDonation);
+    const donation = await Donation.findOne({
+      where: {
+        id: saveDonationResponse.data.data.createDonation,
+      },
+    });
+    assert.equal(donation?.transactionId, transactionId.toLowerCase());
```
Comment (aminlatifi) [inline src/resolvers/donationResolver.test.ts]: Capitalization must be preserved in Solana
```suggestion
    assert.equal(donation?.transactionId, transactionId);
```

File: src/utils/validators/graphqlQueryValidators.ts [hunk 18]
```diff
@@ -19,6 +21,7 @@ const resourcePerDateRegex = new RegExp(
 
 const ethereumWalletAddressRegex = /^0x[a-fA-F0-9]{40}$/;
 const txHashRegex = /^0x[a-fA-F0-9]{64}$/;
+const solanaTxRegex = /^[A-Za-z0-9]{88}$/; // TODO: Is this enough? We are using the signature to fetch transactions
```
Comment (aminlatifi) [inline src/utils/validators/graphqlQueryValidators.ts:24]: The length is usually 43-44 characters, but in theory can be shorter.

File: src/resolvers/donationResolver.test.ts [hunk 19]
```diff
@@ -2246,10 +2295,7 @@ function createDonationTestCases() {
         },
       },
     );
-    assert.equal(
-      saveDonationResponse.data.errors[0].message,
-      errorMessages.INVALID_TRANSACTION_ID,
-    );
+    assert.isOk(saveDonationResponse.data.errors[0]);
```
Comment (aminlatifi) [inline src/resolvers/donationResolver.test.ts]: Why it needs to be changed.

File: src/services/donationService.ts [hunk 20]
```diff
@@ -47,11 +51,19 @@ export const updateDonationPricesAndValues = async (
   project: Project,
   token: Token | null,
   currency: string,
-  priceChainId: number,
+  priceChainId: number | null,
```
Comment (aminlatifi) [inline src/services/donationService.ts]: Let's keep it mandatory and pass zero fo solana

File: test/testUtils.ts [hunk 21]
```diff
@@ -1366,6 +1366,14 @@ export const SEED_DATA = {
         decimals: 18,
       },
     ],
+    solana: [
+      {
+        name: 'Solana native token',
+        symbol: 'SOL',
+        address: '0x0000000000000000000000000000000000000000',
+        decimals: 18,
```
Comment (aminlatifi) [inline test/testUtils.ts]:
```suggestion
        address: '11111111111111111111111111111111',
        decimals: 9,
```

File: src/services/onramper/donationService.ts [hunk 22]
```diff
@@ -140,6 +141,7 @@ export const createFiatDonationFromOnramper = async (
       fiatTransaction.payload.outCurrency,
       priceChainId,
       fiatTransaction.payload.outAmount,
+      ChainType.EVM,
```
Comment (aminlatifi) [inline src/services/onramper/donationService.ts]: Let's make ChainType.EVM the default value and revert changes in other similar places.

File: src/utils/validators/graphqlQueryValidators.ts [hunk 23]
```diff
@@ -75,22 +78,42 @@ export const createDonationQueryValidator = Joi.object({
   amount: Joi.number()?.greater(0).required(),
   transactionId: Joi.when('safeTransactionId', {
     is: null || undefined || '',
```
Comment (aminlatifi) [inline src/utils/validators/graphqlQueryValidators.ts]: The value of `is` property is wrong or does not satisfy the intention of the original writer. The value will always evaluated to ''.

File: src/utils/validators/graphqlQueryValidators.ts [hunk 24]
```diff
@@ -75,22 +78,42 @@ export const createDonationQueryValidator = Joi.object({
...
+        }),
+      Joi.string()
+        .allow(null, '')
+        .pattern(solanaTxRegex, 'Solana Transaction ID')
+        .messages({
+          'string.pattern.base': i18n.__(
+            translationErrorMessagesKeys.INVALID_TRANSACTION_ID,
+          ),
+        }),
+    ),
```
Comment (aminlatifi) [inline src/utils/validators/graphqlQueryValidators.ts]: This alternative is not needed as we don't accept safe transaction on solana

File: test/pre-test-scripts.ts [hunk 25]
```diff
@@ -202,6 +204,16 @@ async function seedTokens() {
     }
     await Token.create(tokenData as Token).save();
   }
+  for (const token of SEED_DATA.TOKENS.solana) {
+    const tokenData = {
+      ...token,
+      networkId: 0,
```
Comment (aminlatifi) [inline test/pre-test-scripts.ts]: Let's use the constant value defined in the src/provider.ts

**Review by aminlatifi — CHANGES_REQUESTED** [review]

File: src/utils/validators/graphqlQueryValidators.ts (hunk 18 above)
Comment (CarlosQ96) [reply src/utils/validators/graphqlQueryValidators.ts:24]: The addresses are 43-44 length.
But transaction ID /TX  is a 88 length ID. 
I think here we can keep this validation

Comment (CarlosQ96) [reply src/utils/validators/graphqlQueryValidators.ts:24]: can refer to here: https://solscan.io/txs

File: src/resolvers/donationResolver.test.ts (hunk 19 above)
Comment (CarlosQ96) [reply src/resolvers/donationResolver.test.ts]: The use of JoI alternative, changes the way error handling works. Now its setting up a generic error. Ill try to setup our old error message, but it might require logic outside the validator.

File: src/utils/networks.ts [hunk 26]
```diff
@@ -2,6 +2,8 @@ import { ethers } from 'ethers';
 import { ChainType } from '../types/network';
 import { PublicKey } from '@solana/web3.js';
 
+export const SOLANA_SYSTEM_ADDRESS = '11111111111111111111111111111111';
```
Comment (aminlatifi) [inline src/utils/networks.ts]: It's Solana system program

File: src/services/onramper/donationService.ts [hunk 27]
```diff
@@ -19,6 +19,7 @@ import {
   updateUserTotalDonated,
   updateUserTotalReceived,
 } from '../userService';
+import { ChainType } from '../../types/network';
```
Comment (aminlatifi) [inline src/services/onramper/donationService.ts]: Is it needed?

File: src/services/donationService.ts [hunk 28]
```diff
@@ -61,21 +75,21 @@ export const updateDonationPricesAndValues = async (
     } else if (token?.cryptoCompareId) {
       const priceUsd = await new CryptoComparePriceAdapter().getTokenPrice({
         symbol: token.cryptoCompareId,
-        networkId: priceChainId,
+        networkId: priceChainId!,
```
Comment (aminlatifi) [inline src/services/donationService.ts]: Is this change needed?

File: src/services/donationService.ts [hunk 29]
```diff
@@ -61,21 +75,21 @@ export const updateDonationPricesAndValues = async (
...
       });
       donation.priceUsd = toFixNumber(priceUsd, 4);
       donation.valueUsd = toFixNumber(donation.amount * priceUsd, 4);
     } else if (token?.coingeckoId) {
       const priceUsd = await new CoingeckoPriceAdapter().getTokenPrice({
         symbol: token.coingeckoId,
-        networkId: priceChainId,
+        networkId: priceChainId!,
```
Comment (aminlatifi) [inline src/services/donationService.ts]: Likewise,

File: src/services/donationService.ts [hunk 30]
```diff
@@ -61,21 +75,21 @@ export const updateDonationPricesAndValues = async (
...
       });
       donation.priceUsd = toFixNumber(priceUsd, 4);
       donation.valueUsd = toFixNumber(donation.amount * priceUsd, 4);
     } else {
       const priceUsd = await new MonoswapPriceAdapter().getTokenPrice({
         symbol: currency,
-        networkId: priceChainId,
+        networkId: priceChainId!,
```
Comment (aminlatifi) [inline src/services/donationService.ts]: likewise,

File: migration/1703044586989-addSolanaToken.ts [hunk 31]
```diff
@@ -0,0 +1,59 @@
...
+export class addSolanaToken1703044586989 implements MigrationInterface {
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.manager.save(
+      Token,
+      seedTokens
+        .filter(token => token.chainType === ChainType.SOLANA)
+        .map(t => {
+          t.address = t.address;
+          return t;
+        }),
```
Comment (aminlatifi) [inline migration/1703044586989-addSolanaToken.ts]: The map doesn't do anything, it must be removed if you didn't have any intention of modification

File: migration/1703044586989-addSolanaToken.ts [hunk 32]
```diff
@@ -0,0 +1,59 @@
...
+    );
+    const tokens = await queryRunner.query(`
+                    SELECT * FROM token
+                    WHERE "chainType" = ${ChainType.SOLANA}
+                    AND "address" = ${SOLANA_SYSTEM_ADDRESS}
```
Comment (aminlatifi) [inline migration/1703044586989-addSolanaToken.ts]: If you want to limit it to the solana system program, this restriction must be applied on saving solana seed token above as well.

File: migration/1703044586989-addSolanaToken.ts [hunk 33]
```diff
@@ -0,0 +1,59 @@
...
+            `);
+    await queryRunner.query(
+      `DELETE FROM organization_tokens_token WHERE "tokenId" IN (${tokens
+        .map(token => token.id)
+        .join(',')})`,
+    );
+    await queryRunner.query(
+      `
+                    DELETE from token
+                    WHERE "networkId" = ${NETWORK_IDS.SOLANA}
```
Comment (aminlatifi) [inline migration/1703044586989-addSolanaToken.ts]: Again, the restriction on the address must be observed here

**Review by aminlatifi — CHANGES_REQUESTED** [review]

File: src/services/onramper/donationService.ts (hunk 27 above)
Comment (CarlosQ96) [reply src/services/onramper/donationService.ts]: removing

File: src/services/donationService.ts (hunk 28 above)
Comment (CarlosQ96) [reply src/services/donationService.ts]: removed

File: src/services/donationService.ts (hunk 29 above)
Comment (CarlosQ96) [reply src/services/donationService.ts]: removing

File: src/services/donationService.ts (hunk 30 above)
Comment (CarlosQ96) [reply src/services/donationService.ts]: removing

**Review by aminlatifi — APPROVED** [review]
LGTM @CarlosQ96, thanks.

**Review by mohammadranjbarz — APPROVED** [review]
Thanks, @CarlosQ96 it needed lots of changes, LGTM
